### PR TITLE
[#16][Chore] Add Unit and UI Test for Login module

### LINKED ISSUE #16: [Chore] Add unit test + ui test for Login Module
## Why

- Login module needs unit tests
 
## Acceptance Criteria

- All Tests passed

### PR DESCRIPTION
**Pull request by nkhanh44**
#16 

## What happened

- Add Unit Tests for Login Screen.
- Add UI Tests for Login Flow.
 
## Insight

- Both Unit tests and UI Tests using Quick & Nimble framework.
- Unit Test covers:
  - `LoginViewModel`
  - `LoginUseCase`
  - `LoginView`
  - `StoreTokenUseCase`

 **Note:** - `HomeView` created in order to fulfill the meaning of the login flow spec in UITest,
## Proof Of Work
Xcode:
Screen Shot 2022-08-09 at 11 33 20
fastlane:
Screen Shot 2022-08-09 at 11 35 10

### CODE REVIEW

File: Survey/Configurations/Keychain/KeychainService.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        try keychain.set(token, key: key.rawValue)
+    }
+
+    func isLoggedIn() throws -> Bool {
+        try keychain.get(Keys.token.rawValue) != nil
+    }
+
+    func removeToken(with key: Keys) throws {
+        try keychain.remove(key.rawValue)
+    }
```
Comment (blyscuit) [inline Survey/Configurations/Keychain/KeychainService.swift]: You can make these functions protocols. When init with UseCases you can declare only the protocol. That way it will be simpler to create Stubs.

File: Survey/Sources/Presentation/Modules/Home/HomeView.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+//  SurveyTests
+//
+//  Created by Khanh on 07/08/2022.
+//  Copyright © 2022 Nimble. All rights reserved.
+//
+
+import Combine
+import SwiftUI
+
+struct HomeView: View {
```
Comment (blyscuit) [inline Survey/Sources/Presentation/Modules/Home/HomeView.swift:12]: The PR and the issue you are working on are only for Unit/UI Tests, not sure why you include HomeView in this PR as well 🤔 . Should you be referencing another issue with HomeView (ideally move these code to another PR to keep at most one issue per PR.).

File: SurveyTests/Domain/Use Case/Login/StoreTokenUseCaseSpec.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    override func spec() {
+
+        var storeTokenUseCase: StoreTokenUseCase!
+        var keychain: KeychainService!
+
+        describe("a StoreTokenUseCase") {
+
+            beforeEach {
+                keychain = KeychainService(service: "")
```
Comment (blyscuit) [inline SurveyTests/Domain/Use Case/Login/StoreTokenUseCaseSpec.swift]: Please create a service with a name for Unit test. But if you're planning to use protocol DI then you might not need service naming at all 😄

File: SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                self.logInUseCase = LoginUseCaseMock()
+                self.storeTokenUseCase = StoreTokenUseCaseMock()
+                viewModel = LoginViewModel(loginUseCase: self.logInUseCase, storeUseCase: self.storeTokenUseCase)
+
+                self.input = LoginViewModel.Input(logInTrigger: logInTrigger.eraseToAnyPublisher())
+                self.output = viewModel.transform(self.input)
+            }
+
+            afterEach {
+                viewModel = nil
```
Comment (blyscuit) [inline SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift]: Wonder why you have to set nil to viewModel? Some bug with the test?

File: SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            context("when login with email and password returns success") {
+
+                it("returns output isLoggedInSuccessfully with true") {
+                    self.input.email = "nkhanh44@nimblehq.co"
```
Comment (blyscuit) [inline SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift]: Please don't use real email/password. Use `example@mail.com` instead.

File: SurveyTests/Presentation/Modules/Login/LoginViewSpec.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+@testable import Survey
+
+final class LoginViewSpec: QuickSpec {
+
+    override func spec() {
+        var loginView: LoginView!
+
+        describe("a LoginView") {
+            beforeEach {
+                loginView = LoginView(viewModel: LoginViewModel(loginUseCase: LoginUseCaseMock(),
```
Comment (blyscuit) [inline SurveyTests/Presentation/Modules/Login/LoginViewSpec.swift]: I think this line break should be
```suggestion
                loginView = LoginView(
                    viewModel: LoginViewModel(
                        loginUseCase: LoginUseCaseMock(),
                        storeUseCase: StoreTokenUseCaseMock()
                    )
                )
```

File: SurveyTests/Sources/Dummy/APIToken+Dummy.swift [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+extension APIToken {
+
+    static let dummy = APIToken(id: "1",
+                                type: "token",
+                                accessToken: "accessToken",
+                                tokenType: "tokenType",
+                                expiresIn: 900,
+                                refreshToken: "refreshToken",
+                                createdAt: 1_888_882)
```
Comment (blyscuit) [inline SurveyTests/Sources/Dummy/APIToken+Dummy.swift]: Please break new line at each brackets. Use auto format after line break.
```suggestion
    static let dummy = APIToken(
                                id: "1",
                                type: "token",
                                accessToken: "accessToken",
                                tokenType: "tokenType",
                                expiresIn: 900,
                                refreshToken: "refreshToken",
                                createdAt: 1_888_882
                                )
```

File: Survey/Sources/Presentation/Modules/Home/HomeView.swift (hunk 2 above)
Comment (nkhanh44) [reply Survey/Sources/Presentation/Modules/Home/HomeView.swift:12]: Simulator Screen Shot - iPhone 12 Pro Max - 2022-08-09 at 17 14 10

Actually since I wrote UITest LoginFlowSpecs, I created HomeView with a button "Home" in order to complete the flow from login to show home screen 😅 . I firstly used a EmptyView(), then I found that's not fulfill the meaning of flow (login -> home), so I created HomeView.

Do my explain answer correctly your concern or do you have any suggestion for me? Thank you 🙏 .

File: SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift (hunk 5 above)
Comment (nkhanh44) [reply SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift]: Thanks, I will fix this 🙇. In case of UITest where I have to use real email/password to run the flow, could i use real email/password? 🙏

Comment (blyscuit) [reply SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift]: Normally we use a mock API (Postman Mock API) and switch the base url with env. This IC don't need to do that so you can use real email/password. Just make sure not to public the repo (your account) 🤣

File: Survey/Sources/Presentation/Modules/Home/HomeView.swift (hunk 2 above)
Comment (blyscuit) [reply Survey/Sources/Presentation/Modules/Home/HomeView.swift:12]: That is fine. You can also create a more minimum View (with out ViewModel) to save on line changes.

Please add in PR comment on why some feature is beyond the issue next time 👍

Comment (nkhanh44) [reply Survey/Sources/Presentation/Modules/Home/HomeView.swift:12]: Thank you, I will do it next time 😄

File: SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift (hunk 4 above)
Comment (nkhanh44) [reply SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift]: There's no bug with the test, I set viewModel to nil after every example since i would like to kinda `deinit` viewModel before the next example. But it seems like a redundant code here since that inits viewModel before each every example 🥲 . I will remove it 💯

File: SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift (hunk 5 above)
Comment (nkhanh44) [reply SurveyTests/Presentation/Modules/Login/LoginViewModelSpec.swift]: Thanks, that's nice information 👍 

> Just make sure not to public the repo (your account)

🥶 🥶 I will notice this.

File: Survey/Sources/Presentation/Modules/Home/HomeViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+import Combine
+import SwiftUI
+
+struct HomeViewModel {}
+
+extension HomeViewModel: ViewModel {
+
+    func transform(_ input: Input) -> Output {
+
```
Comment (blyscuit) [inline Survey/Sources/Presentation/Modules/Home/HomeViewModel.swift]: Please check this line break, I don't think we need one for func inside.

File: SurveyTests/Domain/Use Case/Login/LoginUseCaseSpec.swift [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import Nimble
+import Quick
+@testable import Survey
+
+final class LoginUseCaseSpec: QuickSpec {
+
+    var loginRepository: LoginRepositoryMock!
+
+    override func spec() {
+
```
Comment (blyscuit) [inline SurveyTests/Domain/Use Case/Login/LoginUseCaseSpec.swift]: Please check this line break, I don't think we need one for func inside.

File: SurveyTests/Domain/Use Case/Login/StoreTokenUseCaseSpec.swift [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+import Combine
+import Nimble
+import Quick
+@testable import Survey
+
+final class StoreTokenUseCaseSpec: QuickSpec {
+
+    override func spec() {
+
```
Comment (blyscuit) [inline SurveyTests/Domain/Use Case/Login/StoreTokenUseCaseSpec.swift]: Please check this line break, I don't think we need one for func inside.

File: SurveyTests/Presentation/Modules/Login/LoginViewSpec.swift [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+@testable import Survey
+
+final class LoginViewSpec: QuickSpec {
+
+    override func spec() {
+        var loginView: LoginView!
+
+        describe("a LoginView") {
+            beforeEach {
```
Comment (blyscuit) [inline SurveyTests/Presentation/Modules/Login/LoginViewSpec.swift:22]: Please check this line break, I think we need one after describe 🤔

**Comment by nkhanh44** [comment]
@blyscuit Thanks for your reviewing 🙌 , I fixed all comments and related errors at https://github.com/nkhanh44/ic-survey-ios/pull/17/commits/bae6e2465d17d5651c8925f405c0820074e5bd44, please help me review this again 🙏

**Review by ducbm051291 — APPROVED** [review]
Good job. LGTM 🚀

**Review by blyscuit — APPROVED** [review]
